core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -74,24 +83,33 @@ private void addOrUpdateActivity(PotentialConflictDTO payload, UserAnonymizedDTO
 			}
 		}
 		// Update message only if it is within five seconds to avoid unnecessary cache flushes.
-		else if (now.getTime() - activity.getEndTime().getTime() >= yonaProperties.getAnalysisService().getUpdateSkipWindow())
+		else if (payload.endTime.getTime() - activity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
+				.getUpdateSkipWindow())
 		{
-			updateLastActivity(payload, now, matchingGoal, activity);
+			updateLastActivity(payload, userAnonymized, matchingGoal, activity);
 			cacheService.updateLatestActivityForUser(activity);
 		}
+		else if (payload.startTime.before(activity.getStartTime()))
+		{
+			// can happen if app activity is posted after some network activity is posted during the use of the same app
+			// TODO: remove activities after start time and log a new activity?
+			// or just overwrite start time of the existing activity, creating possible overlap?
```

Comment:
The problem is that we cache the last activity to update it. It may be that there are multiple activities added for network activity while the app is open, e.g. there may be a network activity period A1 at start time t1, another network activity period A2 at start time t2, and the app activity period at start time t0 may span A1 and A2. Updating the start time of A2 to t0 will cause that A1 and A2 overlap.

A purer solution would be to set the start time of A2 to t0 and remove A1. However, then it may be the case that there is already a goal conflict message sent for A1, so that maybe  needs to be relinked to the extended A2.